src/vm/fieldmarshaler.cpp

```diff
@@ -659,7 +661,15 @@ do                                                      \
                 {
                     if (IsStructMarshalable(thNestedType))
                     {
-                        INITFIELDMARSHALER(NFT_NESTEDVALUECLASS, FieldMarshaler_NestedValueClass, (thNestedType.GetMethodTable()));
+                        LONG numBufferElements;
+                        if (IsFixedBuffer(pfwalk->m_MD, pInternalImport, &numBufferElements) && thNestedType.GetMethodTable()->IsBlittable())
```

Comment:
Does it still make sense to use `NFT_NESTEDVALUECLASS` for this now that it is limited to blittable types only? It seems to really just be a really expensive way to do memcpy now.

Reply:
I can revert it back to being a new NFT type. If I do that though, I have to replicate the non-marshaling codepaths that use NFTs to use the same code as `NFT_NESTEDVALUECLASS`. For reference, blittable types contained within non-blittable types still go through the `NestedValueClass` field marshaler.

Reply:
I do not think we need a new NFT type for this. I think that the fix is a lot more complex (and worse perf) than what it needs to be.

Can we fix it by changing `FieldMarshaler_NestedValueClass::NestedValueClassUpdateCLRImpl` to:

```csharp
    MethodTable* pMT = GetMethodTable();

    // would be better to detect this at class load time (that have a nested value
    // class with no layout) but don't have a way to know
    if (!pMT->GetLayoutInfo())
        COMPlusThrow(kArgumentException, IDS_NOLAYOUT_IN_EMBEDDED_VALUECLASS);

    if (pMT->IsBlittable())
    {
        memcpyNoGCRefs((BYTE*)(*ppProtectedCLR) + startoffset, pNative, pMT->GetNativeSize());
    }
    else
    {
        LayoutUpdateCLR( (LPVOID*)ppProtectedCLR,
                             startoffset,
                             GetMethodTable(),
                             (BYTE *)pNative);
    }
}
```

And similar change in `FieldMarshaler_NestedValueClass::NestedValueClassUpdateNativeImpl` ?

Reply:
That should work. I'll try it out.